### ISSUE TITLE
Nesting children/parents in boolean functions

### ISSUE THREAD

**Issue by omstuhler** [vanatteveldt/rsyntax#16]
Hi,

I have noticed that it is currently not possible to nest children/parent functions in boolean queries. Would it be possible at all to implement this or can you think of a clever workaround? Obviously I could just write a new query but I have tons of queries that would have to be duplicated, which is not ideal.

I show an example below (that doesn't work) to illustrate the potential use of making this possible. Basically the query says "grab Peter's actions or actions of things Peter is linked to via "appos".

```
tokens = spacy_parse(c("Peter left.", "My friend Peter left.") , dependency=T)

##### Set key token fo interest
key_token = "Peter"

##### Rule to extract Peter actions in two scenarios
nsubj_act = tquery(OR(AND(token = key_token, relation = "nsubj"),
                      AND(relation = "nsubj", children(token = key_token, relation = "appos"))
                     ),
                   
                   parents(pos = "VERB", 
                           label = "T", custom_fill(BREAK(relation = '*'))
                   )
)

```

**Comment by kasperwelbers** [comment]
The Boolean operators can indeed not contain nested queries. On the one hand the reason for this is technical, as it would be rather hard to implement. Basically, there are two querying components to rsyntax queries. One is the selection of nodes, which is where the Boolean operators are currently designed for. The other is the selection of parents and children. While syntactically it would be possible to also enable OR searches for the latter, it's not an easy feature to add.

The other reason is readability. Although OR conditions on the graph search would require fewer queries, it also makes them much harder to interpret. THe current design of rsyntax is geared towards each query representing a specific type of pattern. 

We do have alternative ways of addressing the type of problem seen in your example. Generally speaking, it often happens that sentences branch in certain ways, and you need to look deeper in the tree to find the token you need. In your case, it seems you want to find the subject-verb relation, but you specifically want the named entity of the subject (in this case Peter). 

What you could do is say: I want to find a verb (pos = VERB), and then I want to recursively (depth larger than 1) search down the tree for a named entity (pos = PROPN). The only tricky thing here is that you specifically want this to be in the 'nsubj' branch. We can to this with the current version, but maybe I can think of a more elegant and flexible way.

What we now can do is something like this: 

```
tq = tquery(label = 'T', pos = 'VERB', fill=F,
                  children(label = 'subject', pos = 'PROPN', depth=2, BREAK(NOT(relation = c('nsubj','appos')))))

```  
In other words: look for the VERB, then look for a child that is a PROPN up to a depth of 2. Also, break the search if your passing over a node that does not have a 'nsubj' or 'appos' relation. 

(notice by the way that I replaced your custom_fill(BREAK(relation='*')) with fill=F. This is because BREAK(relation = '*') basically means "break the fill for ANY possible relation, which is the same as saying don't do fill at all.)

That being said, in case such as this I think it's actually better to use two separate queries. The way rsyntax is designed, there is no harm in making many queries, and there is some merit in having multiple simple queries over one complicated one.

**Comment by omstuhler** [comment]
Thank you for this thorough response! The solution you suggested works perfectly for my problem and I hand't thought of it before.

On a general note, I do think adding flexibility to the query format could be a benefit for the package. The main reason is that 'make another query' is a workable solution for any particular problem - but if you already have 30 queries and you need to duplicate them because of something like this, it's a bit of a headache (luckily the solution suggested allows me to avoid that). From what I can see, nesting parent/children in boolean functions wouldn't fundamentally break with the logic of your query format - but I realize that may be much easier to say on the surface than to implement.

**Comment by kasperwelbers** [comment]
I kind of left this open to push myself to think on this, but time has not been kind on workload. 

In time I do want to consider revamping the query logic, but not yet sure how to approach this. I do not want to tinker with the query format itself, because it is deeply embedded in how query searches are performed (and this is a recursive headache). But I could well imagine a higher level interface for developing queries. Behind the scenes it could then still just generate 30 queries for OR branching, but the reader shouldn't have to do the leg work.

Closing this for now. Feel free to contact me if you have any ideas!
